### DISCUSSION TITLE
Question on Book 1 Chapter 6.7 ray_color function

### DISCUSSION THREAD

**Discussion by VxDxK** [RayTracing/raytracing.github.io#1060] [Q&A] [answered]
Should we unit rec.normal vector in thic code?
`
    hit_record rec;
    if (world.hit(r, 0, infinity, rec)) {
        return 0.5 * (rec.normal + color(1,1,1));
    }
`
Or did I make a mistake earlier somewhere in the Hittable code, becaue **without** this action I`m falling out of color range(from 0 to 1).

**Comment by hollasch** [comment, marked as answer]
Took me a bit to crawl back through the sample code to figure out what's happening. In code listing _The sphere class with normal determination_, you can find `vec3 outward_normal = (rec.p - center) / radius;`. Since `rec.p` is the hit point on the surface of the sphere, `rec.p - center` must be a vector of length `radius`. Thus, dividing by the radius should make this a unit vector. This may be different in your code (you might just have `rec.p - center` only).

Still, if `set_face_normal()` expects given vectors to be of unit length, there's nothing really that guarantees or checks that. Further, the color computation above clearly expects `rec.normal` to be of unit length, so we need to address this.

Spawned issues #1062, #1063.

Thanks for the question.

**Reply by VxDxK** [reply]
Thanks for the answer
